lib/attrsets.nix

```diff
@@ -872,6 +878,62 @@ rec {
     else
       [];
 
+
+  /**
+    Recursively collect sets that verify a given predicate named `pred` from the arbitrary nesting of
+    sets and lists `attrs`. The recursion is stopped when the predicate is verified.
+
+    In other words, it behaves like [`lib.attrsets.collect`](#function-library-lib.attrsets.collect) but also recurses on lists.
+
+    # Inputs
+
+    `pred`
+
+    : Given an attribute's value, determine if recursion should stop.
+
+    `attrs`
+
+    : The arbitrary nesting of sets and lists to recursively collect from.
+
+    # Type
+
+    ```
+    collect :: (AttrSet -> Bool) -> AttrSet -> [x]
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.recursiveCollect` usage example
+
+    ```nix
+    collect (x: x ? outPath)
+      {
+        a = { outPath = "a/"; };
+        b.b = { outPath = "b/"; };
+        c = [
+          { outPath = "c/"; }
+        ];
+      }
+    => [
+      { outPath = "a/"; }
+      { outPath = "b/"; }
+      { outPath = "c/"; }
+    ]
+    ```
+    :::
+  */
+  recursiveCollect =
+  pred:
```

Comment:
@roberth Probably `pred` should also take the `path` just like https://github.com/NixOS/nixpkgs/pull/331384#issuecomment-2308525661